packages/truffle-decode-utils/src/types/types.ts

```diff
@@ -0,0 +1,541 @@
+//note though that values.ts does not include these.
+
+//NOTE: not all of these optional fields are actually implemented. Some are
+//just intended for the future.
+
+import BN from "bn.js";
+import { AstDefinition, AstReferences } from "../ast";
+import { Definition as DefinitionUtils } from "../definition";
+import { Contexts } from "../contexts";
+
+export namespace Types {
+
+  export type Type = UintType | IntType | BoolType | BytesType | AddressType
+    | FixedType | UfixedType | StringType | ArrayType | MappingType | FunctionType
+    | StructType | EnumType | ContractType | MagicType;
+
+  export interface UintType {
+    typeClass: "uint";
+    bits: number;
+  }
+
+  export interface IntType {
+    typeClass: "int";
+    bits: number;
+  }
+
+  export interface BoolType {
+    typeClass: "bool";
+  }
+
+  export type BytesType = BytesTypeStatic | BytesTypeDynamic;
+
+  export interface BytesTypeStatic {
+    typeClass: "bytes";
+    kind: "static";
+    length: number;
+  }
+
+  export interface BytesTypeDynamic {
+    typeClass: "bytes";
+    kind: "dynamic";
+    location?: string;
+  }
+
+  export interface AddressType {
+    typeClass: "address";
+    payable: boolean;
+  }
+
+  export interface StringType {
+    typeClass: "string";
+    location?: string;
+  }
+
+  export interface FixedType {
+    typeClass: "fixed";
+    bits: number;
+    places: number;
+  }
+
+  export interface UfixedType {
+    typeClass: "ufixed";
+    bits: number;
+    places: number;
+  }
+
+  export type ArrayType = ArrayTypeStatic | ArrayTypeDynamic;
+
+  export interface ArrayTypeStatic {
+    typeClass: "array";
+    kind: "static";
+    baseType: Type;
+    length: BN;
+    location?: string;
+  }
+
+  export interface ArrayTypeDynamic {
+    typeClass: "array";
+    kind: "dynamic";
+    baseType: Type;
+    location?: string;
+  }
+
+  export type ElementaryType = UintType | IntType | BoolType | BytesType | FixedType
+    | UfixedType | AddressType | StringType;
+
+  export interface MappingType {
+    typeClass: "mapping";
+    keyType: ElementaryType;
+    valueType: Type;
+    location?: string; //should only ever be storage
+  }
+
+  export interface FunctionType {
+    typeClass: "function";
+    visibility: string; //should be "internal" or "external", not weird
+    //solidity-internal visibilities (but for technical reasons I'm not
+    //putting that in the type requirements)
+    mutability: string; //similarly, should be "nonpayable", "pure", "view",
+    //or "payable"
+    inputParameterTypes: Type[];
+    outputParameterTypes: Type[];
+    //we do not presently support bound functions
+  }
+
+  export type ContractDefinedType = StructType | EnumType;
+  export type UserDefinedType = ContractDefinedType | ContractType;
+
+  export interface StructType {
+    typeClass: "struct";
+    id: number;
+    typeName: string;
+    definingContractName: string;
+    definingContract?: ContractType;
+    memberTypes?: [string, Type][];
```

Comment:
I mean, JS doesn't really have tuples, but TS *does* have support for treating arrays as tuples, else I wouldn't be able to write the type signature above.  And obviously even in JS arrays *do* get used as tuples in functions like `Object.entries`.

IDK.  It's definitely possible I should reevaluate my use of tuples.  So let's see... where do I use tuples currently?  I think basically they make sense for:

A. Stuff that's pretty ad-hoc and not worth declaring a separate type for, and

B. Stuff that has an obvious or conventional ordering.

I think (B) may be sufficient on its own.  (A) is not; can always use an ad-hoc object type if it's just (A).  But if it's (A) and (B), or even just (B), hey, can just use a tuple.

So, where do I actually use tuples currently, and how do they do on these criteria?

1. The `parameters` function you brought up above -- honestly I think this one is fine.  It's only used that one time, and input-then-output is a fairly standard order.

2. Return values of allocation functions, including size, sometimes dynamicity, and the new allocations object -- oof.  These ones smell.  I should replace these with objects.

3. In the `value` field of `MappingValue` -- I think key-then-value is standard enough (see: `Object.entries`) that this is fine.

4. In the `value` field of `StructValue` -- similar.

5. In the `memberTypes` field of `StructType` -- this is the case you're commenting on!  I think this one is fine for the same reason as above?  Regardless it should match (4), because, well, duh.

6. Are there any I missed?

I did make a separate type for decoded parameters to a transaction or event over on `wire-decoder`, rather than using tuples, but that's because there `name` can potentially be empty.  Here that can't happen (I tried just to be sure!).

Anyway yeah we *could* make all these objects rather than tuples, but personally I think (with the exception of (2) above, those I should change) that these are fine as tuples.

Reply:
Hm -- actually I can think of an argument for avoiding tuples in the particular case of structs (4) and (5), but I'm not sure what to think about it.  I'll ask you about it later, don't want to go into it here.  It involves other things.

Reply:
OK I've fixed (2) above.

Reply:
I kind of disagree that key-value pairs are clear on their own. For `memberTypes`, I think something like this would remove all chance of lack of clarity:
```typescript
export interface StructType {
  /* ... */
  memberTypes?: ({ member: string, type: Type })[]
}
```
Although you'd want to have a separate type declaration, of course, and maybe that's why you're objecting.

Anyway, I won't press the issue if you strongly disagree. I'm just on a "everything must be named" kick, I guess.

Reply:
Well, again, let's look at this more generically.  If we were to do this with `StructType`, it wouldn't be `member` and `type`, because I'd want the structure in `StructValue` to be parallel.  I'd want to give them the same names.  Again, the question is of handling key-value pairs in general, not just here.  (I'd also probably call it `name` rather than `member`.)

In any case it wouldn't *need* a separate declaration, it's just a question of which is more convenient.  But key-value pair just seems a sufficiently strong convention to me that it's easier to go with that than to try to remember, oh, was it called `member` or `name` or what?

Reply:
Or hm.  Maybe `name` and `type` would be fine after all.  I guess there's no reason it has to be *exactly* parallel.

Yeah IDK maybe using objects instead of pairs here is fine.